src/julia_gc.c

```diff
@@ -0,0 +1,641 @@
+
+    alloc_size = sizeof(BagHeader) + size;
+    bag = AllocateMasterPointer();
+
+    SizeAllBags             += size;
+
+    /* If the size of an object is zero (such as an empty permutation),
+     * and the header size is a multiple of twice the word size of the
+     * architecture, then the master pointer will actually point past
+     * the allocated area. Because this would result in the object
+     * being freed prematurely, we will allocate at least one extra
+     * byte so that the master pointer actually points to within an
+     * allocated memory area.
+     */
+    if (size == 0)
+      alloc_size++;
+    /* While we use the Boehm GC without the "all interior pointers"
+     * option, stack references to the interior of an object will
+     * still be valid from any reference on the stack. This can lead,
+     * for example, to a 1GB string never being freed if there's an
+     * integer on the stack that happens to also be a reference to
+     * any character inside that string. The garbage collector does
+     * this because after compiler optimizations (especially reduction
+     * in strength) references to the beginning of an object may be
+     * lost.
+     *
+     * However, this is not generally a risk with GAP objects, because
+     * master pointers on the heap will always retain a reference to
+     * the start of the object (or, more precisely, to the first byte
+     * past the header area). Hence, compiler optimizations pose no
+     * actual risk unless the master pointer is destroyed also.
+     *
+     * To avoid the scenario where large objects do not get deallocated,
+     * we therefore use the _ignore_off_page() calls. One caveat here
+     * is that these calls do not use thread-local allocation, making
+     * them somewhat slower. Hence, we only use them for sufficiently
+     * large objects.
+     */
+    BagHeader * header = AllocateBagMemory(type, alloc_size);
+
+    header->type = type;
+    header->flags = 0;
+    header->size = size;
+
+    /* set the masterpointer                                               */
+    SET_PTR_BAG(bag, DATA(header));
+    /* return the identifier of the new bag                                */
+    return bag;
+}
+
+UInt ResizeBag (
+    Bag                 bag,
+    UInt                new_size )
+{
+    UInt                type;           /* type of the bag                 */
+    UInt                flags;
+    UInt                old_size;       /* old size of the bag             */
+    Bag *               src;            /* source in copying               */
+    UInt                alloc_size;
+
+    /* check the size                                                      */
+
+#ifdef TREMBLE_HEAP
+    CollectBags(0,0);
+#endif
+
+    BagHeader * header = BAG_HEADER(bag);
+
+    /* get type and old size of the bag                                    */
+    type     = header->type;
+    flags    = header->flags;
+    old_size = header->size;
```

Comment:
Does Julia GC offer a function similar to `GC_size`? If no, perhaps one could/should add one? (Small allocations seem to come from pools, where the allocation size should thus be known; big allocations seem to store the size in struct `bigval_t` -- but of course I only scratched the surface of that code, so what I just said may very well be just nonsense)